### PR TITLE
DOC, MAINT: Clarify error message of random.randint.

### PR DESCRIPTION
**Pull request by maxwell-aladago**
Improve the error raised by `random.randint` when `low <= 0` and `high` is not given.

The previous behaviour to throw `ValueError: ValueError: low >= high` which is correct but confusing unless you read the source code. I don't think adding that extra check early on will degrade performance.

### CODE REVIEW

**Comment by mattip** [comment]
LGTM. Any reason not to merge?

File: numpy/random/generator.pyx [hunk 1]
```diff
@@ -434,6 +434,8 @@ cdef class Generator:
 
         """
         if high is None:
+            if isinstance(low, (int, np.integer)) and low <= 0:
```
Comment (eric-wieser) [inline numpy/random/generator.pyx]: I'm not really a fan of this type of detection - it doesn't work on things that are coercible to integers.

Could you instead use something like the following?
```python
high_and_low_swapped = high is None:
if high_and_low_swapped:
    high = low
    low = 0

try:
    # the chain of ifs below
except ValueError as e:
    if 'low >= high' == str(e):
        raise ValueError(...) from None
```

File: numpy/random/tests/test_generator_mt19937.py [hunk 2]
```diff
@@ -497,6 +497,7 @@ def test_integers(self):
         actual = random.integers(-99, 99, size=(3, 2))
         desired = np.array([[-80, -56], [41, 37], [-83, -16]])
         assert_array_equal(actual, desired)
+        assert_raises(ValueError, random.integers, 0)
```
Comment (eric-wieser) [inline numpy/random/tests/test_generator_mt19937.py]: Can you use `assert_raises_regex`? This test would pass before your change too.

File: numpy/random/generator.pyx (hunk 1 above)
Comment (maxwell-aladago) [reply numpy/random/generator.pyx]: Thanks @eric-wieser . Just pushed the changes.

File: numpy/random/generator.pyx [hunk 3]
```diff
@@ -451,25 +452,29 @@ cdef class Generator:
...
+                ret = _rand_uint32(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
+            elif key == 'uint16':
+                ret = _rand_uint16(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
+            elif key == 'uint8':
+                ret = _rand_uint8(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
+            elif key == 'bool':
+                ret = _rand_bool(low, high, size, _masked, endpoint, &self._bitgen, self.lock)
+        except ValueError as e:
+            if high_and_low_swapped and 'low >= high' == str(e):
+                e.args = ("low must be greater than 0 when high is not given.", )
```
Comment (eric-wieser) [inline numpy/random/generator.pyx]: I'm not sure if this works correctly in displayed stack traces. What's the trade-off between doing this and raising a brand new exception `from None`?

Comment (maxwell-aladago) [reply numpy/random/generator.pyx]: Yes, the stack trace is good. As an example, the code 
```
import numpy as np
rng = np.random.default_rng()
rng.integers(0)
```
produces the stacktrace 
```
Traceback (most recent call last):
  File "/Users/maxwellaladago/Documents/pub/numpy-max/perms.py", line 4, in <module>
    rng.integers(0)
  File "generator.pyx", line 458, in numpy.random.generator.Generator.integers
  File "bounded_integers.pyx", line 1228, in numpy.random.bounded_integers._rand_int64
ValueError: low must be greater than 0 when high is not given.
```

The problem with the `raise from None` is that it contains the phrase `raise from None` which can be confusing. Additionally, it doesn't capture the trace of the original exception. For example, 

```
import numpy as np
try:
    rng = np.random.default_rng()
    rng.integers(0)
except ValueError as e:
    raise ValueError("just a demo") from None

```
will produce the stacktrace 

```
Traceback (most recent call last):
  File "/Users/maxwellaladago/Documents/pub/numpy-max/perms.py", line 6, in <module>
    raise ValueError("just a demo") from None
ValueError: just a demo
```

Comment (seberg) [reply numpy/random/generator.pyx]: How about just preempting the whole thing and raising the error early on? That seems simpler to parse to me?

**Comment by bashtage** [comment]
It seems strange to have one error check in this file (which is then duplicated).

Why not just expand the current error message to be "low >= high. If high was not provided, then low must be greater than 0.").  This is really simple and avoids extra complexity.  The end-user will get the correct message.

**Comment by maxwell-aladago** [comment]
> It seems strange to have one error check in this file (which is then duplicated).
> 
> Why not just expand the current error message to be `"low >= high. If high was not provided, then low must be greater than 0."`. This is really simple and avoids extra complexity. The end-user will get the correct message.

That involves changing the error message in all the functions inside the try catch block.

**Comment by bashtage** [comment]
@maxwell-aladago There is only 1 function.  They are templated.

**Comment by bashtage** [comment]
The lines are here:

https://github.com/numpy/numpy/blob/master/numpy/random/bounded_integers.pyx.in#L71

and 

https://github.com/numpy/numpy/blob/master/numpy/random/bounded_integers.pyx.in#L145

just append the second part and all is fixed.

**Comment by bashtage** [comment]
Two more places, 

https://github.com/numpy/numpy/blob/master/numpy/random/bounded_integers.pyx.in#L163
https://github.com/numpy/numpy/blob/master/numpy/random/bounded_integers.pyx.in#L289


Could easily refactor to a shared error message (before any of the functions, about here: https://github.com/numpy/numpy/blob/master/numpy/random/bounded_integers.pyx.in#L22,

`low_high_error = 'low {comp} high. If high was not provided, then low must be greater than 0.'`

and then the four `ValueError`s become

`raise ValueError('low_high_error'.format(comp=comp))`.

Increasing complexity for an error message does not seem worth the cost to me.

**Comment by mattip** [comment]
@maxwell-aladago thoughts? I think the general direction is to close this as "too complicated". Maybe there is a better way?

**Comment by bashtage** [comment]
@mattip Given randint is legacy only it might make sense to leave it. If there are changes needed, they should be clarified to apply to `Generator.intergers`

**Pull request closed without merging**